Source/DafnyCompilers/CompilerRewriter.dfy

```diff
@@ -0,0 +1,2215 @@
+        (tr': ExprTransformer)
+      // Given a bottom-up transformer `tr`, return a transformer which applies `tr` in
+      // a bottom-up manner.
+      {
+        var tr': Transformer'<Expr,Expr> := Map_Expr_Transformer'(tr);
+        Map_Expr_Transformer'_Lem(tr);
+        tr'
+      }
+
+      function method {:verify true} {:opaque} Map_Method(m: Method, tr: BottomUpTransformer) :
+        (m': Method)
+        requires Deep.All_Method(m, tr.f.requires)
+        ensures Deep.All_Method(m', tr.post)
+        ensures tr.rel(m.methodBody, m'.methodBody)
+      // Apply a transformer to a method, in a bottom-up manner.
+      {
+        Shallow.Map_Method(m, Map_Expr_Transformer(tr))
+      }
+
+      function method {:verify true} {:opaque} Map_Program(p: Program, tr: BottomUpTransformer) :
+        (p': Program)
+        requires Deep.All_Program(p, tr.f.requires)
+        ensures Deep.All_Program(p', tr.post)
+        ensures tr.rel(p.mainMethod.methodBody, p'.mainMethod.methodBody)
+      // Apply a transformer to a program, in a bottom-up manner.
+      {
+        Shallow.Map_Program(p, Map_Expr_Transformer(tr))
+      }
+    }
+  }
+
+  module Equiv {
+    // This module introduces the relations we use to describe values and expressions
+    // as equivalent, and which we use to state that our compilation passes are sound.
+
+    import DCC = DafnyCompilerCommon
+    import Lib
+    import Lib.Debug
+    import opened DCC.AST
+    import opened Lib.Datatypes
+    import opened Rewriter.BottomUp
+
+    import opened DCC.Predicates
+    import opened Transformer
+    import opened Interp
+    import opened Values
+
+    // TODO: use more
+    datatype Equivs =
+      EQ(eq_value: (WV, WV) -> bool, eq_state: (State, State) -> bool)
+
+    // TODO: not sure it was worth making this opaque
+    predicate {:verify true} {:opaque} GEqCtx(
+      eq_value: (WV,WV) -> bool, ctx: Context, ctx': Context)
+      requires WellFormedContext(ctx)
+      requires WellFormedContext(ctx')
+    {
+      && ctx.Keys == ctx'.Keys
+      && (forall x | x in ctx.Keys :: eq_value(ctx[x], ctx'[x])) 
+    }
+
+    predicate {:verify true} GEqState(
+      eq_value: (WV,WV) -> bool, ctx: State, ctx': State)
+    {
+      GEqCtx(eq_value, ctx.locals, ctx'.locals)
+    }
+
+    function {:verify true} Mk_EqState(eq_value: (WV,WV) -> bool): (State,State) -> bool
+    {
+      (ctx, ctx') => GEqState(eq_value, ctx, ctx')
+    }
+
+    predicate {:verify true} GEqInterpResult<T(0)>(
+      eq_ctx: (State,State) -> bool, eq_value: (T,T) -> bool, res: InterpResult<T>, res': InterpResult<T>)
+    // Interpretation results are equivalent.
+    // "G" stands for "generic".
+    // TODO: be a bit more precise in the error case, especially in case of "out of fuel".
+    {
+      match (res, res') {
+        case (Success(Return(v,ctx)), Success(Return(v',ctx'))) =>
+          && eq_ctx(ctx, ctx')
+          && eq_value(v, v')
```

Comment:
Can we make this a member function in Equivs? (Or in fact two member functions: one for `Result` and one for `Return`)